### PR TITLE
WI00911310 - Add poemGen functionality

### PR DESCRIPTION
**Pull request by max-an-wtg**

### CODE REVIEW

File: preprocess.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+class PoemDataset(Dataset):
+   def __init__(self, poems, char2idx):
+       self.char2idx = char2idx
+       self.unk_idx = char2idx[UNK_CHAR]
+       self.data = []
+       for poem in poems:
+           encoded = [char2idx.get(c, self.unk_idx) for c in poem]
+           if 2 <= len(encoded) <= MAX_LEN:
+               x = encoded[:-1]
```
Comment (hongyuchen-wtg) [inline preprocess.py]: Another choice is adding `<START_CHAR>` and `<END_CHAR>` for each training pair, so we don't even need to end the generation through punctuations but `<END_CHAR>` instead. And once generation ends, we can use a check function to validate the generated text, e.g.
```
poem = ''
currentChar = BEGIN_CHAR
step = 1

while currentChar != END_CHAR and step <= 100:
    step += 1
    poem += genChar
    # feed to model and continue generating

result = poem[1:]
result = result.replace('。', '。\n')

if StandardPoem(result):
    return result

return "Failed to Gen..."

def StandardPoem(poem):
    # split the segment by stop words
    # to check whether the segment number is even number
    # and if want to be stricter: for i in range(0, len(segments, 2)): segment[i] == segment[i+1]

```

File: preprocess.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    char2idx = {char: idx for idx, char in enumerate(vocab)}
+    idx2char = {idx: char for char, idx in char2idx.items()}
+    return vocab, char2idx, idx2char
+
+class PoemDataset(Dataset):
+   def __init__(self, poems, char2idx):
+       self.char2idx = char2idx
+       self.unk_idx = char2idx[UNK_CHAR]
+       self.data = []
+       for poem in poems:
```
Comment (hongyuchen-wtg) [inline preprocess.py]: It'll be better to sort `poems` by length in prior here, so there will be less likely to do padding operation in each batch and helps on faster loss converge.

File: training.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+from tqdm import tqdm
+import torch
+import torch.nn as nn
+from torch.utils.data import DataLoader
+
+from model import CharLSTM
+from preprocess import importData, buildVocab, PoemDataset, collate, PAD_IDX
+
+def train():
+    num_epochs = 999
```
Comment (hongyuchen-wtg) [inline training.py]: too large I think, 30-50 should be fine according to log. Or we can keep it but adding a termination mechanism during training: if avg loss in current epoch is within 1% of the last epoch, end it.

File: training.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+            optimizer.zero_grad()
+
+            output, _ = model(x)
+
+            loss = criterion(output.view(-1, len(vocab)), y.view(-1))
+
+            loss.backward()
+
+            torch.nn.utils.clip_grad_norm_(model.parameters(), 5)
```
Comment (hongyuchen-wtg) [inline training.py:60]: any consideration we choose `5` here?

File: training.py (hunk 3 above)
Comment (hongyuchen-wtg) [reply training.py]: sry 50-100

File: preprocess.py (hunk 2 above)
Comment (hongyuchen-wtg) [reply preprocess.py]: skip this cmt as the `shuffle` is set as `true` in loader

File: generator.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    poems = importData(os.path.join("poemGen", "poetry.txt"))
+    
+    model_path = os.path.join('poemGen', f'model{len(poems)}.pth')
+
+    vocab, char2idx, idx2char = buildVocab(poems)
+
+    model = load_model(
```
Comment (hongyuchen-wtg) [inline generator.py]: no extra lines when inputting parameters

File: preprocess.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+def resourcePath(relative_path):
+
+    if hasattr(sys, '_MEIPASS'):
+
+        return os.path.join(sys._MEIPASS, relative_path)
+
+    return os.path.join(os.path.dirname(__file__), relative_path)
+ 
+DATA_DIR = resourcePath("poemGen")
```
Comment (hongyuchen-wtg) [inline preprocess.py]: when I tried to wrap it with pyinstaller, the resource files path seems have to be adjusted as `~/poemGen/poetry.txt` and `~/poemGen/model43030.pth` because of this line, not sure whether it's on my device only. And `poetry.txt` can be excluded from final result which means we could decouple some settings and only wrap necessary files.
